### PR TITLE
#7622: MV3: Add support for SVG theme icons

### PR DESCRIPTION
**Pull request by fregante**
Ready for review, but not complete.

## What does this PR do?

- Continues from https://github.com/pixiebrix/pixiebrix-extension/pull/7634
- Will eventually fix https://github.com/pixiebrix/pixiebrix-extension/issues/7622

## Checklist

- [ ] Remove hardcoded demo image
- [ ] Deploy our own proxy
- [ ] Update tests
- [x] Designate a primary reviewer: @grahamlangford

### CODE REVIEW

**Comment by fregante** [comment]
Current status:

- all online solutions expect this issue to be solved via `postMessage`, which is not available for MV3 `offscreen` documents
- `runtime.sendMessage`  does not appear to support "transferable" objects, so I can't pass around Blob, ImageData, ImageBitmap
- those types aren't exactly "serializable"
- probably the only way I can do this is to:
	1. pass the `url` to the offscreen document
	2. return a raw array of pixels in the offscreen document (via `getImageData().data`)
	3. use `OffscreenCanvas` again to paint it in the background worker
	4. use `getImageData()` to get the proper `ImageData` object again

**Comment by fregante** [comment]
Can we just setup a proxy? This appears to have too many small issues to be maintainable. If we spun up a small server/lambda/vercel we could pass an SVG URL and receive a PNG, preserving most of the existing code here

```diff
+ if (url.endsWith('svg')) {
+  url = `https://svg-to-png.pixiebrix.com/?width=32&height=32&url=${url}`
+ }
const { data } = await axios.get<Blob>(url, { responseType: "blob" });
return await blobToImageData(data);
```

The proxy itself can use any off-the-shelves SVG-to-PNG converter. Example:

- source: https://github.com/GewoonJaap/svg-to-png-cf-worker
- live URL: https://svg-to-png.mrproper.dev/https://docs.tandoor.dev/logo_color.svg

We'll need it until it's natively supported:

- https://github.com/w3c/webextensions/issues/462

**Comment by fregante** [comment]
Most of the code improvements were extracted to a non-MV3-specific PR https://github.com/pixiebrix/pixiebrix-extension/pull/7935

File: src/background/setToolbarIconFromTheme.ts [hunk 1]
```diff
@@ -25,6 +25,8 @@ export default async function setToolbarIconFromTheme({
   toolbarIcon,
   themeName,
 }: Pick<ThemeAssets, "logo" | "toolbarIcon" | "themeName">) {
+  toolbarIcon =
+    "https://upload.wikimedia.org/wikipedia/commons/f/fd/Ghostscript_Tiger.svg";
   if (toolbarIcon) {
```
Comment (fregante) [inline src/background/setToolbarIconFromTheme.ts:30]: Hard-coded demo image. This image is SVG and lacks a height/width attribute, which further complicates things without a proxy.

File: src/utils/canvasUtils.ts [hunk 2]
```diff
@@ -18,32 +18,34 @@
...
-  // TODO: URL.createObjectURL() and Image() is not available in service workers
-  // https://groups.google.com/a/chromium.org/g/chromium-extensions/c/u0NH7L3v9L4
-  // https://github.com/pixiebrix/pixiebrix-extension/issues/7622
-  const url = URL.createObjectURL(blob);
-  const image = new Image();
-  image.src = url;
-  await image.decode();
-  // `createImageBitmap` will fail on SVGs that lack width/height attributes, so we must use <img>
-  return image;
+  return blobToImageData(blob, width, height);
```
Comment (fregante) [inline src/utils/canvasUtils.ts:48]: Boom. This is all that's needed to fully support SVGs both in MV2 and MV3.

File: src/utils/canvasUtils.ts [hunk 3]
```diff
@@ -18,32 +18,34 @@
 import axios from "axios";
 import resizeToFit from "intrinsic-scale";
 
+async function loadAsBlob(url: string): Promise<Blob> {
+  const response = await axios.get<Blob>(url, { responseType: "blob" });
+  return response.data;
+}
+
+function getSvgProxyUrl(url: string): string {
+  return `https://svg-to-png.mrproper.dev/${url}`;
```
Comment (fregante) [inline src/utils/canvasUtils.ts:27]: Rust-based demo proxy working on CloudFlare Workers. We need to deploy our own before merging this PR

https://github.com/GewoonJaap/svg-to-png-cf-worker

**Comment by grahamlangford** [comment]
> Can we just setup a proxy? This appears to have too many small issues to be maintainable. If we spun up a small server/lambda/vercel we could pass an SVG URL and receive a PNG, preserving most of the existing code here
> 
> ```diff
> + if (url.endsWith('svg')) {
> +  url = `https://svg-to-png.pixiebrix.com/?width=32&height=32&url=${url}`
> + }
> const { data } = await axios.get<Blob>(url, { responseType: "blob" });
> return await blobToImageData(data);
> ```
> 
> The proxy itself can use any off-the-shelves SVG-to-PNG converter. Example:
> 
> * source: https://github.com/GewoonJaap/svg-to-png-cf-worker
> * live URL: https://svg-to-png.mrproper.dev/https://docs.tandoor.dev/logo_color.svg
> 
> We'll need it until it's natively supported:
> 
> * [Proposal: Add Support for Dynamic SVG Icons in `browser.action.setIcon()` w3c/webextensions#462](https://github.com/w3c/webextensions/issues/462)

It's unfortunate Heroku doesn't support serverless functions. Let me discuss with @johnnymetz what the best approach would be for us. I don't see why we couldn't. 

Alternatively, we might just setup an endpoint on App.

**Comment by fregante** [comment]
We also have an account on Vercel already so we could set that up. CloudFlare might be even safer (more generous free plan, as well as a repo ready to be deployed)

**Comment by grahamlangford** [comment]
@johnnymetz will create an App ticket to look into this. We're going to pursue the cf worker solution initially.

**Comment by johnnymetz** [comment]
Here's the issue: https://github.com/pixiebrix/pixiebrix-extension/issues/7982

**Comment by twschiller** [comment]
Closing out this PR because we're going to be creating/hosting the PNG based icon: https://www.notion.so/pixiebrix/Support-custom-branding-MV3-browser-action-icon-f43f5cf790ed4776b15e23638f5ce580?pvs=4

**Comment by fregante** [comment]
If we don't support SVG here, we still need to apply most changes in this PR

**Pull request closed without merging**